### PR TITLE
Doc: move prodn for filtered_import and import_categories

### PR DESCRIPTION
**Pull request by SkySkimmer**
Moved to be closer to their respective explanations

### CODE REVIEW

File: doc/sphinx/language/core/modules.rst [hunk 1]
```diff
@@ -197,12 +197,6 @@ are now available through the dot notation.
 
 .. cmd:: Import {? @import_categories } {+ @filtered_import }
 
-   .. insertprodn import_categories filtered_import
-
-   .. prodn::
-      import_categories ::= {? - } ( {+, @qualid } )
-      filtered_import ::= @qualid {? ( {+, @qualid {? ( .. ) } } ) }
-
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:205]: Keep this.

File: doc/sphinx/language/core/modules.rst [hunk 2]
```diff
@@ -244,6 +238,11 @@ are now available through the dot notation.
 
          Check B.T.
 
+   .. insertprodn filtered_import filtered_import
+
+   .. prodn::
+      filtered_import ::= @qualid {? ( {+, @qualid {? ( .. ) } } ) }
+
```
Comment (jfehrle) [inline doc/sphinx/language/core/modules.rst:245]:
```suggestion
   :n:`@qualid {? ( {+, @qualid {? ( .. ) } } ) }`
```

**Review by jfehrle — COMMENTED** [review]
The convention we've followed is to give the productions in the insertprodn right after the command or tactic that needs them.  Then text explaining each part of the syntax, perhaps repeating some of the syntax using ```:n:`...` ```.    I think we should stick with that. 
 The destruct tactic is a good example of this format.

The examples for this command should be appear after all the parameters have been described.  The text can hyperlink to them.

As I expect you don't want to take on all of this, a reasonable compromise is repeat only the rhs of each production as I indicated.

**Comment by SkySkimmer** [comment]
>The convention we've followed is to give the productions in the insertprodn right after the command or tactic that needs them. 

Is that really true? For instance `qualid` is further down in this file.

Regardless that compromise isn't really satisfying to me, my main goal is to have `#grammar-token-import_categories` (linked whenever import_categories appears in a production) go to the explanation for import categories instead of the top of the explanation of Import.

**Comment by jfehrle** [comment]
It is as true as we can make it.  We can't repeat nonterminal definitions such as `qualid` because, as you know, they are the target of hyperlinks.

**Comment by SkySkimmer** [comment]
Indeed IMO the situation here is the same as qualid, I want to move the target of the hyperlink.

**Comment by SkySkimmer** [comment]
I still think having the link target go next to the explanation is better than your convention.

**Comment by Zimmi48** [comment]
@jfehrle is right that this is the convention that we have been following thus far, but I agree with @SkySkimmer that this:

> Regardless that compromise isn't really satisfying to me, my main goal is to have `#grammar-token-import_categories` (linked whenever import_categories appears in a production) go to the explanation for import categories instead of the top of the explanation of Import.

is a compelling reason for moving things. In fact, I would consider any case where following a link then requires scrolling to find the corresponding documentation to be a documentation bug.

In this case, we can consider two solutions:

- move the grammar production to the place where it is documented, like this PR proposes
- or, at the very least, add a link directly under the grammar rules to the place where they are documented (would require following two links)

I would tend to favor the second option only when there are very few links to the grammar production, and that the command under which the grammar rules are located can be considered sufficiently important (compared to the sources of the links) to justify this.

But in this case, a further issue is that the "import categories" are identifiers in a specific list, which appears much below.

---

Independently of these considerations, the way `Import` is documented is quite unsatisfying:


In the above, there is no way to know which `qualid` is referred to (is it the one in `import_categories` or `filtered_import`?). When the production rules are moved, it becomes even more important to fix this. We usually do it, by writing:

```rst
If :token:`qualid` (in :n:`@filtered_import`) denotes a valid basic module (i.e. its module type is a
```

**Comment by jfehrle** [comment]
> In the above, there is no way to know which `qualid` is referred to (is it the one in `import_categories` or `filtered_import`?). 

Another way to handle this is to add subscripts on `qualid` in the productions and the text.  In the following example, the subscripts appear as `one_term__from` (with two underscores).  Hyperlinks ignore the subscripts.  You have to add them by hand in the rst.  If the tactic or production is modified, `doc_gram_rst` loses the subscripts.

image

**Review by proux01 — APPROVED** [review]

**Comment by proux01** [comment]
So there seems to be a clear majority considering this an improvement over the current situation, even if this departs a bit from a convention we are not stricly following anyway (see for instance https://coq.inria.fr/doc/V8.20.0/refman/user-extensions/syntax-extensions.html#grammar-token-syntax_modifier ). Let's merge, this has waited way too much already.
@coqbot merge now

**Pull request merged**